Add capacitance calculation function for parallel plate capacitor with support for multiple inputs

dnxjay opened the pull request:
This pull request introduces a Python function that calculates the capacitance of a parallel plate capacitor based on the area of the plates and the distance between them. It includes  function capacitance(area, distance) to compute capacitance, Constant Definition, Support for Single and Multiple Inputs, Testing, Lambda and Map for Efficient Batch Processing

dnxjay committed:
Implemented basic capacitance calculation with lambda and map

- Added a function to calculate the capacitance of a parallel plate capacitor using the formula C = epsilon_0 * (A / d).
- Defined the permittivity constant (epsilon_0) and wrote a function 'capacitance' that takes area and distance as parameters.
- Extended the functionality using functional programming concepts:
   - Used a lambda function to compute capacitance for multiple sets of areas and distances.
   - Applied the 'map' function to apply the lambda function to lists of areas and distances.
- The code is currently printing capacitances for both single inputs and lists of multiple areas and distances.

dnxjay committed:
made some adjustments to add the Lambda and Map:

dnxjay committed:
Added unit tests for capacitance function

- Created a new file 'test_capacitance.py' to implement unit tests for the capacitance function.
- The tests validate the correctness of the capacitance function using assert statements.
- Checked if the calculated capacitance matches the expected value based on given areas and distances.
- Ensured that the test prints "All tests passed" when successful, verifying the correct behavior of the function.

dnxjay committed:
Implemented capacitance calculation with functional programming and unit tests

dnxjay committed:
Added capacitance() function to calculate the capacitance of a parallel plate capacitor.
Introduced batch_capacitances() using list comprehension for calculating capacitances for multiple areas and distances.
Implemented resistance() function to calculate the resistance of a wire based on length, area, and resistivity.
Added batch_resistances() using map() and lambda for batch resistance calculation.
Created unit tests for both capacitance and resistance functions using assert statements.
Structured the code into two separate modules (capacitance.py and resistance.py) as per the project requirements.
Ensured that functional programming concepts (list comprehension, map(), and lambda) were used in the implementation.

dnxjay committed:
Merge branch 'main' into main

laserlab commented:
@jkblc please review asap

jkblc reviewed (changes requested):
Please address the linting problems

laserlab commented on capacitance_module/__pycache__/capacitance.cpython-312.pyc:
no binary files, please remove

laserlab commented on capacitance_module/capacitance.py:
```diff
@@ -0,0 +1,33 @@
+#ance.py (extended with lambda and map)
```
what is ance.py?

laserlab commented on capacitance_module/capacitance.py:
```diff
@@ -0,0 +1,33 @@
+    Calculate the capacitance of a parallel plate capacitor.
+
+    Parameters:
+    area (float): Area of the plate in square meters.
```
do not specific a data type in python in the do string if that is not actually checked or necessary

laserlab commented on capacitance_module/capacitance.py:
```diff
@@ -0,0 +1,33 @@
+    return epsilon_0 * (area / distance)
+
+# Test the function with example values
+if __name__ == "__main__":
```
Testing should be done in the unit test not in a main loop

dnxjay replied:
can you explain this part with more detail? I believe some of my issues to having a functional code at the moment is due to this

laserlab commented on capacitance_module/resistance.py:
```diff
@@ -0,0 +1,29 @@
+    Calculate the resistance of a wire.
+
+    Parameters:
+    length (float): Length of the wire in meters.
```
see above

laserlab commented on capacitance_module/resistance.py:
```diff
@@ -0,0 +1,29 @@
+    Calculate resistances for a list of lengths, areas, and resistivities using map and lambda.
+
+    Parameters:
+    lengths (list of float): List of wire lengths.
```
list is useful, but not float

laserlab commented on capacitance_module/test_resistance.py:
docstrings missing

laserlab commented on capacitance_module/test_capacitance.py:
docstrings missing

laserlab reviewed (changes requested):
The actual homework task of using your electrodynamic function inside a function is not implemented.

laserlab commented:
> Please address the linting problems. Otherwise looks good

Please check my review comments for issues the linter doesn’t catch

laserlab commented on capacitance_module/test_capacitance.py:
```diff
@@ -0,0 +1,28 @@
+    areas = [1.0, 2.0, 0.5]
+    distances = [0.01, 0.02, 0.005]
+    expected_batch = [epsilon_0 * (A / d) for A, d in zip(areas, distances)]
+    assert batch_capacitances(areas, distances) == expected_batch, "Batch capacitance test failed"
```
Linter complains about, but in general you should actually run your tests …

dnxjay committed:
Added documentation for Capacitance of a Parallel Plate Capacitor

- Explained the physical concept of capacitance and the formula for a parallel plate capacitor.
- Included the impact of dielectric materials on capacitance.
- Provided examples of capacitance calculations with and without a dielectric.
- Discussed parasitic capacitance and its effect in high-frequency circuits.
- Added information on how to calculate total capacitance for capacitors in series and parallel.
- Added references for further reading.

dnxjay committed:
Merge remote-tracking branch 'origin/main'

dnxjay committed:
Delete

dnxjay committed:
delete

dnxjay committed:
Added functions for Capacitance of a Parallel Plate Capacitor

- Implemented `calculate_capacitance` to compute the capacitance of a parallel plate capacitor based on the area of the plates and the distance between them.
- Added `calculate_capacitance_with_dielectric` to handle capacitance calculations when a dielectric material is present.
- Implemented `calculate_total_capacitance_series` to compute total capacitance for capacitors connected in series.
- Implemented `calculate_total_capacitance_parallel` to compute total capacitance for capacitors connected in parallel.
- Added `calculate_parasitic_capacitance` to compute the parasitic capacitance between components or wires.
also Added unit tests for Capacitance of a Parallel Plate Capacitor functions

- Added `test_basic_capacitance` to validate basic capacitance calculation for parallel plate capacitors.
- Added `test_capacitance_with_dielectric` to test capacitance calculation with a dielectric material.
- Added `test_total_capacitance_series` to check total capacitance for capacitors connected in series.
- Added `test_total_capacitance_parallel` to check total capacitance for capacitors connected in parallel.
- Added `test_parasitic_capacitance` to validate parasitic capacitance calculations for wires.
- Included `test_very_small_area_and_distance` to test edge cases with very small values for area and distance.
- All tests implemented using the `unittest` framework.

dnxjay committed:
filename change

jkblc reviewed (changes requested):
You still need to use your electrodynamic function inside a function

dnxjay committed:
Linting issues

dnxjay committed:
linting/ code productivity

jkblc commented on capacitance_module/__pycache__/capacitance.cpython-312.pyc:
Please remove the binary file

jkblc reviewed (changes requested):
Looks better. Please remove the binary file

jkblc commented on capacitance_module/capacitance.py:
You still need to use your electrodynamics function inside another function. That is an important part of the homework

dnxjay replied:
Im confused about that, the code does include functions that use principles from electrodynamics inside other functions. The functions calculate_capacitance, calculate_capacitance_with_dielectric, and calculate_parasitic_capacitance all rely on electrodynamic formulas involving permittivity, area and distance between plates to calculate capacitance

jkblc replied:
Maybe @laserlab can answer this, but each "def" in your code is a function. We are not directly talking about math functions.

jkblc replied:
We discussed it in class today while talking about not assigning lambda functions to a variable

dnxjay committed:
removed  binary file

laserlab commented:
@jkblc Is correct no use of ED function as an argument in another function. Not merging.

dnxjay commented:
since mine wasnt merged, i dont get any credit for the work ive done???

laserlab commented:
you do get points for the work, just not the merge points
